bliss/surveys/dc2.py

```diff
@@ -1,55 +1,217 @@
+import collections.abc  # noqa: WPS301
 import math
+import multiprocessing
 import pathlib
 import random
 
 import numpy as np
 import pandas as pd
 import torch
 from astropy.io import fits
+from astropy.io.fits import Header
 from astropy.wcs import WCS
 from einops import rearrange
-from torch.utils.data import DataLoader
+from torch.utils.data import DataLoader, Dataset
 
 from bliss.catalog import FullCatalog
 from bliss.surveys.survey import Survey
 
 
+def from_wcs_header_str_to_wcs(wcs_header_str: str):
+    return WCS(Header.fromstring(wcs_header_str))
+
+
+def map_nested_dicts(cur_dict, func):
+    if isinstance(cur_dict, collections.abc.Mapping):
+        return {k: map_nested_dicts(v, func) for k, v in cur_dict.items()}
+
+    return func(cur_dict)
+
+
+def generate_split_file(image_index, object_attributes):
+    split_count = 0
+    image_list, bg_list, wcs, wcs_header_str = read_frame_for_band(
+        object_attributes["image_files"],
+        object_attributes["bg_files"],
+        image_index,
+        object_attributes["n_bands"],
+        object_attributes["image_lim"],
+    )
+    image = np.stack(image_list)
+    bg = np.stack(bg_list)
+
+    plocs_lim = image[0].shape
+    height = plocs_lim[0]
+    width = plocs_lim[1]
+    full_cat, psf_params, _ = Dc2FullCatalog.from_file(
+        object_attributes["cat_path"], wcs, height, width, object_attributes["bands"]
+    )
+    tile_cat = full_cat.to_tile_catalog(4, 5).get_brightest_sources_per_tile()
+
+    tile_dict = tile_cat.to_dict()
+
+    tile_dict["locs"] = rearrange(tile_cat["locs"], "1 h w nh nw -> h w nh nw")
+    tile_dict["n_sources"] = rearrange(tile_cat["n_sources"], "1 h w -> h w")
+    tile_dict["source_type"] = rearrange(tile_cat["source_type"], "1 h w nh nw -> h w nh nw")
+    tile_dict["galaxy_fluxes"] = rearrange(tile_cat["galaxy_fluxes"], "1 h w nh nw -> h w nh nw")
+    tile_dict["galaxy_params"] = rearrange(tile_cat["galaxy_params"], "1 h w nh nw -> h w nh nw")
+    tile_dict["star_fluxes"] = rearrange(tile_cat["star_fluxes"], "1 h w nh nw -> h w nh nw")
+    tile_dict["star_log_fluxes"] = rearrange(
+        tile_cat["star_log_fluxes"], "1 h w nh nw -> h w nh nw"
```

Comment:
A for loop here might make the code more concise. e.g.,
```
for k, v in tile_dict.items():
    if k != "n_sources":
        tile_dict[k] = rearrange(v, "1 h w nh nw -> h w nh nw")
```

Reply:
It'd be even better if we didn't need to squeeze these tensors...is that possible? (Perhaps by using them below in the unsqueezed format.)

Reply:
fix it in commit a521a09, and we will remove the squeeze in next PR